xdem/terrain.py

```diff
@@ -0,0 +1,232 @@
+"""Terrain attribute calculations, such as the slope, aspect etc."""
+from __future__ import annotations
+
+from typing import Sized, overload
+
+import numpy as np
+
+import xdem.spatial_tools
+
+
+@overload
+def get_terrain_attribute(
+    dem: np.ndarray | np.ma.masked_array,
+    attribute: str,
+    resolution: tuple[float, float] | float,
+    degrees: bool = True,
+    hillshade_altitude: float = 45.0,
+    hillshade_azimuth: float = 315.0,
+    hillshade_z_factor: float = 1.0,
+) -> np.ndarray:
+    ...
+
+
+@overload
```

Comment:
I looked at the documentation, but I'm still not sure I understand completely what `@overload` does here exactly! I see the types of `attribute` and output are different, one is a list and the other a single value

Reply:
It's a slightly clunky, but more or less functional way of describing return types that depend on the input arguments.

Let's say I have a function that provides word counts for one or multiple words:
```python

def word_count(words: str | list[str]) -> int | list[int]:
    if isinstance(words, str):
         words = [words]
    counts = list(map(len, words))

    return counts if len(counts) > 1 else counts[0]
```
The "problem" with this function is that linters and type checkers can't know whether the function will return an integer or a list of integers, so `@overload` is there to provide that information. If `@overload` is not used, the user has to treat the return type as either a list or an integer every time, which can get very tedious!

```python
from typing import overload

@overload
def word_count(words: str) -> int: ...

@overload
def word_count(words: list[str]) -> list[int]: ...

# here goes the function definition above
```
Does that make sense?

Reply:
Yes, perfect sense! Thanks a lot for the example :)